docs:Basic - Lifting State Up (#6)
reactjs/id.reactjs.org: content/docs/lifting-state-up.md

---
id: lifting-state-up

> title: Mengangkat Ke Atas

Reviewer: `Mengangkat State ke Atas` atau `Memindahkan State ke Atas` sepertinya lebih cocok mas 😄

  - "docs/flux-todo-list.html"
---

> Seringkali, beberapa komponen perlu mencerminkan perubahan data yang sama. Kami menyarankan mengangkat `state` bersama ke modul utama terdekat. Mari kita lihat bagaimana ini bekerja dalam aksi.

Reviewer 2: > Seringkali, beberapa komponen perlu mencerminkan perubahan data yang sama.

Lebih baik diganti seperti ini:

> Seringkali, beberapa komponen perlu mencerminkan perubahan dari data yang sama.

Karena maksud sebenarnya adalah untuk mencerminkan data yang sama, yang sedang berubah. Intinya di "data", bukan di "perubahan data"-nya. Perlu lebih diperjelas meskipun maknanya hampir sama.

---

> Kami menyarankan mengangkat `state` bersama ke modul utama terdekat.

Lebih baik diganti seperti ini:

> Kami menyarankan untuk memindah `state` yang digunakan bersama tersebut ke modul _parent_ terdekat yang dimiliki komponen-komponen terkait.

Di bagian ini, kami akan membuat kalkulator suhu yang menghitung apakah air akan mendidih pada suhu tertentu.

> Kami akan memulai dengan komponen bernama `BoilingVerdict`. Ini menerima `celsius` suhu sebagai `props`, dan mencetak apakah itu cukup untuk merebus air:

Reviewer: `We` di sini sepertinya mengacu pada `Kita`.

```
Ini menerima `celsius` suhu sebagai `props`, dan mencetak apakah itu cukup untuk merebus air:
```

sepertinya lebih cocok: 
```
Komponen tersebut akan menerima suhu `celsius` sebagai *prop*, dan menentukan apakah suhu cukup untuk membuat air mendidih
```

Reviewer 2: Setuju dengan review mas @regalius, jangan lupa gunakan _props_, bukan _prop_ karena di https://github.com/reactjs/id.reactjs.org/wiki/Glosarium sudah ditentukan.

Reviewer 2: Maaf, saya ingin meralat review saya, betul menggunakan _prop_ karena kalimat aslinya juga begitu. Yang perlu dirubah adalah penggunaan code block untuk _prop_. Gunakan _emphasis_ untuk term yang tidak bisa ditranslate, bukan `code block`.

```js{3,5}
function BoilingVerdict(props) {
  if (props.celsius >= 100) {

>     return <p>Air akan mendidih.</p>;

Reviewer: Teks di dalam kode sebaiknya jangan di translasi dulu mas 😄

  }
  return <p>Air tidak akan mendidih.</p>;
}
```

> Selanjutnya, kita akan membuat komponen bernama `Calculator`. Ini membuat `<input>` yang memungkinkan anda memasukan suhu, dan menyimpan nilainya di `this.state.temperature`.

Reviewer 2: > Ini membuat `<input>` yang memungkinkan anda memasukan suhu, dan menyimpan nilainya di `this.state.temperature`.

Lebih baik diganti seperti ini:

> Komponen tersebut menampilkan `<input>` yang memungkinkan Anda untuk memasukkan nominal suhu, dan menyimpan nilainya di `this.state.temperature`.

Perlu diperhatikan bahwa ada kesalahan di _translation_ yang mas buat:
- `anda` seharusnya `Anda`
- `memasukan` seharusnya `memasukkan`
- `renders` tidak dapat diartikan ke `membuat`. Gunakan arti yang lebih cocok `menampilkan` atau jika tidak memungkinkan gunakan kata aslinya `render` menggunakan _emphasis_.

> Selain itu, ini membuat `BoilingVerdict` untuk nilai input saat ini.

Reviewer 2: Sama, ubah kalimat ini menjadi:

> Selain itu, komponen tersebut menampilkan `BoilingVerdict` yang menggunakan nilai masukan `celcius` yang terbaru.

Coba cari penyampaian kalimat yang lebih efektif dari sekedar men-_translate_ dari Bahasa Inggris.

> }

Reviewer: String & Komen di dalam blok kode boleh di translate mas 😄

    const temperature = this.state.temperature;
    return (
      <fieldset>

>         <legend>Masukan suhu dalam Celcius:</legend>

Reviewer 2: Ubah `masukan` ke `masukkan`.

}
```

> [**Cobalah di CodePen**](https://codepen.io/gaearon/pen/ZXeOBm?editors=0010)

Reviewer 2: Kalau menggunakan kata `Cobalah`, sebaiknya diikuti dengan kata benda-nya. Seperti `Cobalah komponen ini di ..`. Jika tidak menggunakan kata benda, lebih baik gunakan kata `Coba` saja.

> [**Cobalah di CodePen**](https://codepen.io/gaearon/pen/ZXeOBm?editors=0010)

Reviewer 2: `Cobalah di` -> `Coba di` supaya seragam dengan PR lain.

## Menambahkan Input Kedua {#adding-a-second-input}

> Persyaratan baru kami adalah bahwa, selain input Celcius, kami memberikan input Fahrenheit dan mereka tetap sinkron.

Reviewer: `Our` di dalam kalimat ini sepertinya konteksnya `Kita`

`input` -> `masukan`
`output` -> `keluaran`

Reviewer 2: Tambahan dari saya:
- bedakan `masukan` dan `masukkan`. Untuk arti kata `input` gunakan `masukan` karena kata tersebut merupakan kata benda. Untuk kata kerja yang diambil dari kata `masuk`, gunakan `masukkan`.

> Persyaratan baru kami adalah bahwa, selain input Celcius, kami memberikan input Fahrenheit dan mereka tetap sinkron.

Reviewer 2: > Persyaratan baru kami adalah bahwa, selain input Celcius, kami memberikan input Fahrenheit dan mereka tetap sinkron.

Lebih baik diganti seperti ini:

> Syarat baru dari kita adalah selain masukan Celcius, kita juga memberikan masukan Fahrenheit, dan kedua masukan tersebut harus tetap sinkron.

> Kita bisa memulai dengan mengekstraksi komponen `TemperatureInput` dari `Calculator`. Kami akan menambahkan `scale` *prop* baru untuk dapat berupa `"c"` atau `"f"`:

Reviewer: `Kami akan menambahkan` sepertinya konteksnya `Kita akan menambahkan`

> Kita bisa memulai dengan mengekstraksi komponen `TemperatureInput` dari `Calculator`. Kami akan menambahkan `scale` *prop* baru untuk dapat berupa `"c"` atau `"f"`:

Reviewer 2: `mengekstraksi` dalam penggunaan Bahasa Indonesia pada umumnya mempunyai makna yang kurang _pas_ jika disandingkan dengan penggunaan `extract` dalam Bahasa Inggris yang lebih fleksibel. Coba gunakan kata lain seperti `mengeluarkan` atau `memisahkan`.

> Kita bisa memulai dengan mengeluarkan komponen `TemperatureInput` dari `Calculator`. Kita akan menambahkan `scale` *prop* baru untuk dapat berupa `"c"` atau `"f"`:

Reviewer: ```suggestion
Kita dapat memulai dengan mengeluarkan komponen `TemperatureInput` dari `Calculator`. Kita akan menambahkan *prop* `scale` baru pada komponen tersebut yang dapat berupa `"c"` atau `"f"`:
```

What do you think? 😄

}
```

> Kita sekarang dapat mengubah `Calculator` untuk membuat dua input suhu terpisah:

Reviewer: render -> `me-*render*`

> Kita sekarang dapat mengubah `Calculator` untuk membuat dua input suhu terpisah:

Reviewer 2: ``Sekarang kita dapat mengubah `Calculator` untuk me-_render_ dua masukan suhu terpisah:``

}
```

[**Cobalah di CodePen**](https://codepen.io/gaearon/pen/jGBryx?editors=0010)

> Kami memiliki dua input sekarang, tetapi ketika anda memasukkan suhu di salah satu dari mereka, yang lain tidak memperbarui. Ini bertentangan dengan persyaratan kami: kami ingin tetap menyinkronkannya.

Reviewer: `We have two inputs` sepertinya konteksnya `Kita sekarang telah memiliki dua input`

`This contradicts our requirement: we want to keep them in sync.`

sepertinya lebih cocok:
`Hal ini bertentangan dengan persyaratan kita: kita ingin agar mereka tetap sinkron`

> Kami memiliki dua input sekarang, tetapi ketika anda memasukkan suhu di salah satu dari mereka, yang lain tidak memperbarui. Ini bertentangan dengan persyaratan kami: kami ingin tetap menyinkronkannya.

Reviewer 2: - Di tempat lain Anda menggunakan `temperatur` untuk mengganti kata `temperature`, sekarang berganti `suhu`, lebih baik gunakan kata yang sama
- `Anda`
- `input` -> `masukan`
- `di salah satu dari mereka` -> `di salah satunya`
- Kalau dirasa kata `memperbarui` dalam konteks kalimat tersebut dirasa kurang cocok, silakan diganti ke kata aslinya, menggunakan _emphasis_ `ter-_update_`

> Kami juga tidak dapat menampilkan `BoilingVerdict` dari `Calculator`. `Calculator` tidak tahu suhu saat ini karena tersembunyi di dalam `TemperatureInput`.

Reviewer: `We also can't display` sepertinya `Kita juga tidak dapat menampilkan`

Reviewer 2: `` `Calculator` tidak tahu suhu saat ini karena suhu tersebut tersembunyi di dalam `TemperatureInput`.``

> ## Penulisan Fungsi Konversi {#writing-conversion-functions}

Reviewer: `Menulis Fungsi Konversi` karena `Writing` dalam konteks ini adalah kata kerja aktif

> Pertama, kita akan menulis dua fungsi untuk mengkonversi dari Celcius ke Fahrenheit dan kembali:

Reviewer: `Pertama-tama, kita akan menulis dua fungsi untuk mengkonversi suhu dari Celsius ke Fahrenheit dan sebaliknya`

}
```

> Dua fungsi ini mengkonversi angka. Kami akan menulis fungsi lain yang menggunakan fungsi *string* `temperature` dan fungsi konverter sebagai argumen dan mengembalikan string. Kami akan menggunakannya untuk menghitung nilai dari suatu input bersarakan input lainnya.

Reviewer: `We will write another function...` -> `Kita akan menulis fungsi...`

Reviewer 2: - `Dua fungsi ini mengkonversi angka.` -> `Dua fungsi ini dapat mengkonversi angka.` karena makna tersebut lebih cocok jika disandingkan dengan kalimat aslinya.
- ``Kami akan menulis fungsi lain yang menggunakan fungsi *string* `temperature` dan fungsi konverter sebagai argumen dan mengembalikan string.``
-> ``Kita akan menulis fungsi lain yang menggunakan _string_ `temperature` dan fungsi konverter sebagai argumen dan menghasilkan sebuah _string_``

> Dua fungsi ini dapat mengkonversi angka. Kita akan menulis fungsi lain yang menggunakan fungsi *string* `temperature` dan fungsi konverter sebagai argumen dan mengembalikan string. Kita akan menggunakannya untuk menghitung nilai dari suatu masukan bersarakan masukan lainnya.

Reviewer: ```suggestion
Dua fungsi ini akan mengkonversi angka. Kita akan menulis fungsi lain yang menerima *string* `temperature` dan sebuah fungsi konverter sebagai argumen dan mengembalikan sebuah *string*. Kita akan menggunakannya untuk menghitung nilai dari suatu masukan berdasarkan masukan lainnya.
```

> Ini mengembalikan string kosong pada `temperature` yang tidak valid, dan itu membuat *output* dibulatkan ke tempat desimal ketiga:

Reviewer 2: - ``Ini mengembalikan string kosong pada `temperature` yang tidak valid, dan itu membuat *output* dibulatkan ke tempat desimal ketiga:``
-> ``Fungsi tersebut menghasilkan _string_ kosong pada `temperature` yang tidak valid, dan juga _output_ dari fungsi tersebut akan dibulatkan ke tempat desimal ketiga.``

}
```

> Misalnya, `tryConvert('abc', toCelsius)` mengembalikan string kosong, dan `tryConvert('10.22', toFahrenheit)` mengembalikan `'50.396'`.

Reviewer 2: - `mengembalikan` -> `menghasilkan`
- `string` -> `_string_`

> ## Mengangkat Ke Atas {#lifting-state-up}

Reviewer: Disamakan dengan judul jadi `Memindahkan State ke Atas` aja mas

  render() {
    const temperature = this.state.temperature;
    // ...
```

Namun, kita ingin kedua masukan selaras satu sama lain. Ketika kita memperbarui masukan Celcius, masukan Fahrenheit harus mencerminkan suhu yang dikonversi, dan sebaliknya.

> Dalam React, *state* bersama dicapai dengan memindahkannya ke modul utama terdekat dari komponen yang membutuhkannya. Ini disebut "mengangkat keadaan". Kita akan menghapus *state* lokal dari `TemperatureInput` dan memindahkannya ke `Calculator` sebagai gantinya.

Reviewer: ```suggestion
Dalam React, *state* bersama dicapai dengan memindahkannya ke komponen induk bersama terdekat dari komponen yang membutuhkannya. Ini disebut "memindahkan *state* ke atas". Kita akan menghapus *state* lokal dari `TemperatureInput` dan memindahkannya ke `Calculator` sebagai gantinya.
```

> Jika `Calculator` memiliki *state* bersama, itu menjadi "sumber kebenaran" untuk suhu saat ini di kedua masukan. Ini dapat menginstruksikan mereka berdua untuk memiliki nilai yang konsisten satu sama lain. Karena kedua `props` pada komponen `TemperatureInput` berasal dari koponen `Calculator` induk yang sama, kedua masukan akan selalu sinkron.

Reviewer: ```suggestion
Jika `Calculator` memiliki *state* bersama, *state* tersebut akan menjadi "sumber kebenaran" untuk suhu saat ini di kedua masukan. Ini dapat menginstruksikan kedua input untuk memiliki nilai yang konsisten satu sama lain. Karena kedua `props` pada komponen `TemperatureInput` berasal dari komponen `Calculator` induk yang sama, kedua masukan akan selalu sinkron.
```

    // ...
```

> Kita tahu bahwa [*props* hanya bisa dibaca](/docs/components-and-props.html#props-are-read-only). Ketika `temperature` berada di *state* lokal, `TemperatureInput` hanya bisa memanggil `this.setState()` untuk mengubahnya. Namun, sekarang bahwa `temperature` berasal dari induknya sebagai `props`, `TemperatureInput` tidak memiliki kendali atasnya.

Reviewer: ```suggestion
Kita tahu bahwa [*props* bersifat *read-only*](/docs/components-and-props.html#props-are-read-only). Ketika `temperature` berada di *state* lokal, `TemperatureInput` bisa memanggil `this.setState()` untuk mengubahnya. Namun, karena sekarang `temperature` berasal dari induknya sebagai sebuah `prop`, `TemperatureInput` tidak memiliki kendali atasnya.
```

    // ...
```

>Catatan:
>
>Tidak ada arti khusus untuk salah satu `temperature` atau `onTemperatureChange` nama *prop* dalam komponen khusus. Kita bisa memanggil mereka apapun, seperti memberi nama mereka `value` dan `onChange` yang merupakan konvensi umum.

> `onTemperatureChange` *prop* akan diberikan bersama-sama dengan `temperature` *prop* oleh induk komponen `Calculator`. Ini akan menangani perubahan dengan memodifikasi *state* lokal sendiri, sehingga menampilkan ulang kedua masukan dengan nilai - nilai baru. Kita akan segera melihat implementasi `Calculator` yang baru.

Reviewer: ```suggestion
*Prop* `onTemperatureChange` akan diberikan bersama-sama dengan  *prop* `temperature` oleh komponen induk `Calculator`. Fungsi ini akan menangani perubahan dengan memodifikasi *state* lokalnya sendiri, sehingga menampilkan ulang kedua masukan dengan nilai-nilai baru. Kita akan segera melihat implementasi `Calculator` yang baru.
```

> Sebelum melihat ke perubahan di `Calculator`, mari kita rekap perubahan kita ke komponen `TemperatureInput`. Kita telah menghapus *state* lokal dari situ, dan alih-alih membaca `this.state.temperature`, kita sekarang membaca `this.props.temperature`. Alih-alih memanggil `this.setState()` saat kita ingin melakukan perubahan, kita sekarang memanggil `this.props.onTemperatureChange()`, yang akan disediakan oleh `Calculator`:

Reviewer: ```suggestion
Sebelum melihat perubahan di `Calculator`, mari kita rekap perubahan kita ke komponen `TemperatureInput`. Kita telah menghapus *state* lokal darinya, dan alih-alih membaca `this.state.temperature`, kita sekarang membaca `this.props.temperature`. Alih-alih memanggil `this.setState()` saat kita ingin melakukan perubahan, kita sekarang memanggil `this.props.onTemperatureChange()`, yang akan disediakan oleh `Calculator`:
```

}
```

> Jika nanti kita mengubah bidang Fahrenheit menjadi 212, state pada `Calculator` akan menjadi:

Reviewer: ```suggestion
Jika nanti kita mengubah kolom Fahrenheit menjadi 212, state pada `Calculator` akan menjadi:
```

}
```

[**Coba di CodePen**](https://codepen.io/gaearon/pen/WZpxpz?editors=0010)

> Sekarang, tidak peduli apabila masukan berubah, `this.state.temperature` dan `this.state.scale` dalam `Calculator` akan ter-*update*. Salah satu masukan mendapatkan nilai apa adanya, sehingga setiap masukan pengguna dipertahankan, dan nilai masukan lainnya selalu dihitung ulang berdasarkan masukan.

Reviewer: ```suggestion
Sekarang, tidak peduli apabila masukan berubah, `this.state.temperature` dan `this.state.scale` dalam `Calculator` akan diperbarui. Salah satu masukan mendapatkan nilai apa adanya, sehingga setiap masukan pengguna dipertahankan, dan nilai masukan lainnya selalu dihitung ulang berdasarkan masukan.
```

Mari kita rekap apa yang terjadi ketika Anda mengubah masukan:

> * React memanggil fungsi yang spesifik `onChange` pada DOM `<input>`. Dalam kasus kita, ini adalah metode `handleChange` dalam komponen `TemperatureInput`.

Reviewer: ```suggestion
* React memanggil fungsi yang disebut `onChange` pada DOM `<input>`. Dalam kasus kita, ini adalah metode `handleChange` dalam komponen `TemperatureInput`.
```

* Metode `handleChange` dalam komponen `TemperatureInput` memanggil `this.props.onTemperatureChange()` dengan nilai baru. Itu merupakan *props*, termasuk `onTemperatureChange`, disediakan oleh komponen induknya, `Calculator`.

> * Ketika sebelumnya di *render*, `Calculator` telah ditentukan bahwa `onTemperatureChange` dari Celcius `TemperatureInput` adalah metode `Calculator` `handleCelsiusChange`, dan `onTemperatureChange` dari Fahrenheit `TemperatureInput` adalah metode `Calculator` `handleFahrenheitChange`. Jadi salah satu dari dua metode `Calculator` dipanggil tergantung pada masukan yang kita ubah.

Reviewer: ```suggestion
* Ketika sebelumnya di-*render*, `Calculator` telah menentukan bahwa `onTemperatureChange` dari `TemperatureInput` Celcius adalah metode `handleCelsiusChange` milik `Calculator`, dan `onTemperatureChange` dari `TemperatureInput` Fahrenheit adalah metode `handleFahrenheitChange` milik `Calculator`. Jadi salah satu dari dua metode `Calculator` dipanggil tergantung pada masukan yang kita ubah.
```

* Di dalam metode ini, komponen `Calculator` meminta React untuk mengulang *render* dengan memanggil `this.setState()` dengan nilai masukan baru dan skala masukan saat ini yang baru saja kita ubah.

> * React memanggil komponen `Calculator` metode `render` untuk mempelajari seperti apa tampilan UI. Nilai kedua masukan dihitung berdasarkan suhu saat ini dan skala aktif. Konversi suhu dilakukan di sini.

Reviewer: ```suggestion
* React memanggil komponen `Calculator` metode `render` untuk mempelajari seperti apa tampilan antarmuka pengguna. Nilai kedua masukan dihitung berdasarkan suhu saat ini dan skala aktif. Konversi suhu dilakukan di sini.
```

> * React memanggil metode `render` pada komponen individu `TemperatureInput` dengan *props* baru yang ditentukan oleh `Calculator`. Ia mempelajari seperti apa tampilan UI mereka.

Reviewer: ```suggestion
* React memanggil metode `render` pada komponen `TemperatureInput` individual dengan *props* baru yang ditentukan oleh `Calculator`. Mreka kemduain mempelajari seperti apa tampilan antarmuka pengguna mereka.
```

> * React memanggil metode `render` pada komponen `TemperatureInput` individual dengan *props* baru yang ditentukan oleh `Calculator`. Mreka kemduain mempelajari seperti apa tampilan antarmuka pengguna mereka.

Reviewer 3: Typo: Mreka -> Mereka

* React memanggil metode `render` pada komponen `BoilingVerdict`, melewati suhu dalam Celcius sebagai *props*.

> * React DOM memperbarui dengan vonis didih dan untuk mencocokan nilai masukan yang diinginkan. Masukan yang baru saja kita ubah menerima nilainya saat ini, dan masukan lainnya diperbarui kesuhu setelah konversi.

Reviewer: ```suggestion
* React DOM memperbarui dengan vonis didih dan untuk mencocokan nilai masukan yang diinginkan. Masukan yang baru saja kita ubah menerima nilainya saat ini, dan masukan lainnya diperbarui ke suhu setelah konversi.
```

> * React DOM memperbarui dengan vonis didih dan untuk mencocokan nilai masukan yang diinginkan. Masukan yang baru saja kita ubah menerima nilainya saat ini, dan masukan lainnya diperbarui ke suhu setelah konversi.

Reviewer 3: "React DOM memperbarui dengan vonis didih" -> "React DOM memperbaharui nilai titik didih"

Setiap pembaruan melewati langkah yang sama sehingga masukan tetap sinkron.

## Pelajaran yang Dipetik {#lessons-learned}

> Seharusnya ada "sumber kebenaran" tunggal untuk setiap data yang berubah dalam aplikasi React. Biasanya, *state* ditambahkan pertama kali ke komponen yang membutuhkannya untuk *rendering*. Kemudian, jika komponen lain juga membutuhkannya, Anda dapat mengangkat ke modul utama terdekat. Alih-alih mencoba menyinkronkan keadaan antara komponen yang berbeda, Anda harus mengandalkan [aliran data dari atas kebawah](/docs/state-and-lifecycle.html#the-data-flows-down).

Reviewer: ```suggestion
Seharusnya ada "sumber kebenaran" tunggal untuk setiap data yang berubah dalam aplikasi React. Biasanya, *state* ditambahkan pertama kali ke komponen yang membutuhkannya untuk di-*render*. Kemudian, jika komponen lain juga membutuhkannya, Anda dapat memindahkan *state* ke komponen induk bersama terdekat. Alih-alih mencoba menyinkronkan *state* antara komponen yang berbeda, Anda harus mengandalkan [aliran data dari atas kebawah](/docs/state-and-lifecycle.html#the-data-flows-down).
```

> Pengangkatan *state* melibatkan penulisan lebih banyak kode "*boilerplate*" daripada pendekatan pengikatan dua arah, tetapi sebagai manfaat, dibutuhkan lebih sedikit pekerjaan untuk menemukan dan mengisolasi *bug*. Karena keadaan apa pun "hidup" di beberapa komponen dan komponen itu sendiri dapat mengubahnya, area permukaan untuk bug sangat berkurang. Selain itu, Anda dapat menerapkan logika khusus apa pun untuk menolak atau mengubah masukan pengguna.

Reviewer: ```suggestion
Pengangkatan *state* melibatkan penulisan lebih banyak kode "*boilerplate*" daripada pendekatan pengikatan dua arah, tetapi sebagai manfaat, dibutuhkan lebih sedikit pekerjaan untuk menemukan dan mengisolasi *bug*. Karena tiap *state* akan "hidup" di sebuah komponen dan hanya komponen itu sendiri yang dapat mengubahnya, area permukaan untuk *bug* akan sangat berkurang. Selain itu, Anda dapat menerapkan logika khusus apa pun untuk menolak atau mengubah masukan pengguna.
```

> Jika sesuatu dapat diturunkan dari *props* atau *state*, itu mungkin tidak boleh di *state*. Sebagai contoh, alih-alih menyimpan keduanya `celsiusValue` dan `fahrenheitValue`, kita menyimpan hanya yang terakhir diubah `temperature` dan itu *scale*. Nilai masukan lain selalu dapat dihitung dari mereka dalam metode `render()`. Ini memungkinkan kita menghapus atau menerapkan pembulatan ke bidang lain tanpa kehilangan ketepatan dalam masukan pengguna.

Reviewer: ```suggestion
Jika sesuatu dapat diturunkan dari *props* atau *state*, hal itu mungkin tidak sebaiknya berada di *state*. Sebagai contoh, alih-alih menyimpan keduanya `celsiusValue` dan `fahrenheitValue`, kita menyimpan hanya `temperature` yang terakhir diubah dan `scale`-nya. Nilai dari masukan lain selalu dapat dihitung dari kedua nilai tersebut dalam metode `render()`. Ini memungkinkan kita menghapus atau menerapkan pembulatan ke masukan lain tanpa kehilangan ketepatan pada masukan pengguna.
```

> Ketika Anda melihat sesuatu yang salah di UI, Anda dapat menggunakan [Alat Pengembang React](https://github.com/facebook/react-devtools) untuk memeriksa *props* dan geser *tree* sampai Anda menemukan komponen yang bertanggung jawab untuk memperbarui *state*. Ini memungkinkan Anda melacak *bug* ke sumbernya:

Reviewer: ```suggestion
Ketika Anda melihat sesuatu yang salah di antarmuka pengguna, Anda dapat menggunakan [Alat Pengembang React](https://github.com/facebook/react-devtools) untuk memeriksa *props* dan menelusuri *tree* komponen anda ke atas sampai Anda menemukan komponen yang bertanggung jawab untuk memperbarui *state*. Ini memungkinkan Anda melacak *bug* ke sumbernya:
```
